### DISCUSSION TITLE
Less Verbose TryParse

### DISCUSSION THREAD

**Discussion by andrewdavidjamesburke** [dotnet/runtime#80962] [General]
# Summary
A less verbose `TryParse` method, parsing directly from the string object into its relative type

## Current Implementation
**Fig 1.1**
```
string s = "3.14159265359"
System.Single.TryParse(s, out float f);
```

## Proposed Implementation
**Fig 1.2**
```
string s = "3.14159265359"
s.TryParse(out float f);
```
## Workaround
**Fig 1.3**
```
//FloatExtensions.cs

public static class FloatExtensions
{
    public static void TryParse(this string s, out float f)
    {
        System.Single.TryParse(s, out f);
    }
}
```

## Motivation
Our team regularly uses variations of the `TryParse` extension method (**Fig 1.3**) for various string operations. 
We have been using (with no observable overhead), and an improvement to workflow of **up to 10 seconds per usage**.
We feel that the entire C# community could benefit from this improvement.

## Setback
The `TryParse` method has been used in its current state in several revisions of the language. This new change would require users to alter their workflow when using `TryParse` methods.
As such, it is also **proposed that a larger number of up votes** than the usual be required for this new change to be adopted, should this proposal be accepted as a candidate.

**Comment by 333fred** [comment]
If you're proposing a new extension method on `String`, the appropriate repo for that is https://github.com/dotnet/runtime. This repo is for C# language design, and I'm not seeing any new syntax proposed here, just a new method.

**Comment by FaustVX** [comment]
You can write your own extension method to include every `TryParse` methods
```C#
public static bool TryParse<T>(this string s, [MaybeNullWhen(false)] out T f)
    where T : System.ISpanParsable<T>
    => T.TryParse(s, null, out f);
```
And use it like that
```C#
if ("3.141592".TryParse(out double pi))
    Console.WriteLine($"pi = {pi}");
```

[sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIAYAEqCMA6AEQEsBDAcygHsBnYIgY2rwGFK4BTAQShIBsBPakWoBuALAAoSUQBmmABQAiAMx4cAFhwBWAJwAmRXgAqYfgAUSYau3mUIwTHDsAjXu0wAHIgEpvkzAHY+Lg68gAkil6YALyYAN5eAL6K3uJSEgD0GQBivJQkwACiAB7A7FBClBV4jJKSqMpBAGzYepi5+UWl5ZUVknH+gQ3NmM6UlLyYJuaW1gA8RgB88sAAFsJBWNQwmADaALIk/M7sAHIQvLwA6qvl8jJ81t4Auph2DkaYMn4SgX+YAHdbmB3J8QEE8ABJADKHhIUAsVhIrnYC0Wg3+MUWU2MpkR1nk20wUAuvB27y+qUkiSAA==)

**Reply by andrewdavidjamesburke** [reply]
> You can write your own extension method to include every `TryParse` methods
> 
> ```cs
> public static bool TryParse<T>(this string s, [MaybeNullWhen(false)] out T f)
>     where T : System.ISpanParsable<T>
>     => T.TryParse(s, null, out f);
> ```
> 
> And use it like that
> 
> ```cs
> if ("3.141592".TryParse(out double pi))
>     Console.WriteLine($"pi = {pi}");
> ```
> 
> [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIAYAEqCMA6AEQEsBDAcygHsBnYIgY2rwGFK4BTAQShIBsBPakWoBuALAAoSUQBmmABQAiAMx4cAFhwBWAJwAmRXgAqYfgAUSYau3mUIwTHDsAjXu0wAHIgEpvkzAHY+Lg68gAkil6YALyYAN5eAL6K3uJSEgD0GQBivJQkwACiAB7A7FBClBV4jJKSqMpBAGzYepi5+UWl5ZUVknH+gQ3NmM6UlLyYJuaW1gA8RgB88sAAFsJBWNQwmADaALIk/M7sAHIQvLwA6qvl8jJ81t4Auph2DkaYMn4SgX+YAHdbmB3J8QEE8ABJADKHhIUAsVhIrnYC0Wg3+MUWU2MpkR1nk20wUAuvB27y+qUkiSAA==)

Simply amazing.

**Comment by koszeggy** [comment]
Maybe it's not just me but I happen to have such [publicly available](https://www.nuget.org/packages/KGySoft.CoreLibraries) extensions. The generic [`TryParse<T>(string, out T)`](https://docs.kgysoft.net/corelibraries/html/M_KGySoft_CoreLibraries_StringExtensions_TryParse__1_1.htm) and [`TryParse(ReadOnlySpan<char>, out T)`](https://docs.kgysoft.net/corelibraries/html/M_KGySoft_CoreLibraries_StringExtensions_Parse__1.htm) extension methods are [optimized](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/c573aa0001a82382fc8ce81808f5364aecfddc32/KGySoft.CoreLibraries/CoreLibraries/_Extensions/StringExtensions.Parser.cs#L291) for known value types.
